My solution

koder95 opened the pull request.

koder95 committed:
Add JPA annotations to model classes

Annotated Actor, Country, and Movie classes with JPA annotations for entity mapping, including @Entity, @Id, @GeneratedValue, @ManyToOne, and @ManyToMany. Updated pom.xml to include hibernate-core and mysql-connector-j dependencies to support JPA and database connectivity.

koder95 committed:
Create DataProcessingException.java

koder95 committed:
Update HibernateUtil.java

koder95 committed:
Update Main.java

koder95 committed:
Implement entity persistence in DAO

Added generic create and find methods to AbstractDao for entity persistence and retrieval. Updated ActorDaoImpl, CountryDaoImpl, and MovieDaoImpl to use these methods, ensuring related entities are persisted as needed before saving the main entity. This improves consistency and reduces code duplication in DAO implementations.

koder95 committed:
Refactor service implementations to use DAOs

Updated ActorServiceImpl, CountryServiceImpl, and MovieServiceImpl to use their respective DAO interfaces for data access. Added constructors for dependency injection of DAOs and SessionFactory. Implemented add and get methods in ActorServiceImpl and MovieServiceImpl to delegate to the DAO layer.

koder95 committed:
Create hibernate.cfg.xml

mateacademy-ai-mentor commented on src/main/java/mate/academy/hibernate/relations/dao/impl/AbstractDao.java:
```diff
@@ -1,11 +1,41 @@
+                session.close();
+            }
+        }
+    }
+
+    protected <T> Optional<T> find(Class<T> entityType, Long id) {
+        try (Session session = factory.openSession()) {
+            return Optional.ofNullable(session.find(entityType, id));
+        }
+    }
```
This method is missing a `catch` block. According to checklist item #9, "All exceptions caught in the DAO layer must be wrapped and re-thrown as the custom `DataProcessingException`". Please add a `catch` block to handle potential runtime exceptions from Hibernate and wrap them in your custom exception.

mateacademy-ai-mentor commented on src/main/java/mate/academy/hibernate/relations/service/impl/CountryServiceImpl.java:
```diff
@@ -1,9 +1,23 @@
+
     @Override
     public Country add(Country country) {
         return null;
```
This method is not implemented. It should call the `add` method on the `dao` object to persist the `country` entity.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

This is a strong submission with a well-implemented structure for the entities and DAO/Service layers. However, a couple of critical issues need to be addressed before I can approve it.

First, the `CountryServiceImpl` is incomplete. The `add` and `get` methods in `src/main/java/mate/academy/hibernate/relations/service/impl/CountryServiceImpl.java` are not implemented and currently return `null`, which causes the application to fail. Please complete these methods by delegating the calls to the corresponding `countryDao` methods.

Second, the `find` method in `src/main/java/mate/academy/hibernate/relations/dao/impl/AbstractDao.java` is missing required exception handling. According to the requirements [CHECKLIST ITEM #9], all exceptions in the DAO layer must be caught and wrapped in a `DataProcessingException`. Please add a `catch` block to handle potential runtime exceptions from Hibernate.

Once these two issues are resolved, your solution will be ready!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

koder95 committed:
Implement CountryService methods

Implemented the add and get methods in CountryServiceImpl to delegate to the DAO. Enhanced AbstractDao's find method to throw a DataProcessingException on errors for better error reporting.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Excellent work on this submission! I'm happy to approve your solution. You have successfully addressed all the points from the previous review, resulting in a robust and complete project. The implementation of the DAO and service layers is solid, the exception handling is now correct, and all components work together seamlessly. The project now fully meets all the requirements. Well done!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
